server/src/deployment/utils.ts

```diff
@@ -17,4 +25,53 @@ export function arcsKeyFor(fingerprint: string):string {
    return ARCS_KEY_PREFIX + fingerprint;
 }
 
+
+/**
+ * Helper to wait a certain number of times for a GCP operation to finish.
+ * @param func an async function that performs a GCP operation and returns a Promise<T>
+ * @param waitCond an async function that takes a T and returns true or false if the operation is completed.
+ * @param retries how many times to execute waitCond before giving up.
+ * @param maxTimeout maximum amount of timeouts allowed.
+ * @return the result from executing func
+ */
+export async function waitForGcp<T>(func: () => PromiseLike<[T]>, waitCond:(result:T) => PromiseLike<boolean>,
+                                    retries = 10, maxTimeout = 30000):Promise<T> {
+
+  let waiting = true;
+  const timeout = maxTimeout / retries;
+  try {
+    const [result] = await Promise.race([func(), delay<[T]>(timeout)]);
+
+    while (waiting && retries-- > 0) {
+      try {
+        // if the timeout finishes first, it will reject and skip this
+        waiting = !await waitCond(result);
+      } catch (e) {
+        if (!(e instanceof Timeout)) {
+          return Promise.reject(e);
+        }
+      }
+      if (waiting) {
+        try {
+          await delay(timeout);
+        } catch (e) {
+        }
```

Comment:
Probably want to log something if we hit an exception?

Reply:
I realize it's confusing, but the Promise rejection is expected. The delay() function returns a Promise that is never resolved, only rejected after it's timeout expires. 

This is used in Promise.race(promise, delay) to detect a timeout where the first promises take too long. It is reused in this fragment to actually wait. I could make a version of delay() that resolves after timeout to avoid the try/catch trick, or hide this try/catch block inside of a helper which I'll do for now.